contracts/SinglePlayerCommit.sol

```diff
@@ -51,213 +61,470 @@ contract SinglePlayerCommit is Ownable {
+    /// @param committer address of the creator of the committer to process
+    /// @dev Process commitment by lookup based on address, checking metrics, state and updating balances
     function processCommitment(address committer) public {
         console.log("Processing commitment");
-        Commitment memory commitment = commitments[committer];
+        Commitment storage commitment = commitments[committer];
 
-        // check if commitment has ended
         require(commitment.endTime < block.timestamp, "SPC::processCommitment - commitment is still active");
+        require(commitment.endTime < commitment.lastActivityUpdate, "SPC::processCommitment - update activity");
 
-        bool met = commitment.met;
-        uint256 stake = commitment.stake;
+        commitment.met = commitment.reportedValue > commitment.goalValue;
 
-        // "delete" the expired commitment
-        commitments[committer].exists = false;
+        if (!commitment.met) {
+            _slashFunds(commitment.stake, committer);
+        } 
+        
+        commitment.exists = false;
+        emit CommitmentEnded(committer, commitment.met, commitment.stake);
+    }
 
-        uint256 penalty;
+    /// @notice Enables control of processing own commitment. For instance when completed.
+    /// @dev Process commitment by lookup msg.sender, checking metrics, state and updating balances
+    function processCommitmentUser() public {
+        console.log("Processing commitment");
+        Commitment storage commitment = commitments[msg.sender];
 
-        if (met) {
-            penalty = 0;
-        } else {
-            penalty = stake;
-            // remove from committer's balance
-            _changeCommitterBalance(penalty, false);
-        }
+        commitment.met = commitment.reportedValue > commitment.goalValue;
 
-        emit CommitmentEnded(committer, met, penalty);
+        if (!commitment.met) {
+            _slashFunds(commitment.stake, msg.sender);
+        } 
+        
+        commitment.exists = false;
+        emit CommitmentEnded(msg.sender, commitment.met, commitment.stake);
     }
 
+    //TODO state change after transfer is not recommended
+    /// @notice Contract owner can withdraw funds not owned by committers. E.g. slashed from failed commitments
+    /// @param amount Amount of <token> to withdraw
+    /// @dev Check amount against slashedBalance, transfer amount and update slashedBalance
     function ownerWithdraw(uint256 amount) public onlyOwner returns (bool) {
-        uint256 available = token.balanceOf(address(this)).sub(committerBalance);
+        console.log("Received call for owner withdrawal for amount %s", amount);
 
-        require(amount <= available, "SPC::ownerWithdraw - not enough available balance");
+        require(amount <= slashedBalance, "SPC::ownerWithdraw - not enough available balance");
         require(token.transfer(msg.sender, amount), "SPC::ownerWithdraw - token transfer failed");
```

Comment:
do the transfer after updating `slashedBalance` (re-entry protection)

Reply:
Done